File: collectors/memory.py
import psutil
import json
from datetime import datetime, timezone

def collect(hostname):
    ts = datetime.now(timezone.utc)
    vm = psutil.virtual_memory()
    swap = psutil.swap_memory()
    # pagefile on Windows via WMI fallback via psutil
    try:
        # Use psutil swap as pagefile approximation
        page_total = swap.total
        page_used = swap.used
    except:
        page_total = page_used = None

    raw = {"virtual": vm._asdict(), "swap": swap._asdict()}
    row = (
        ts, hostname,
        int(vm.total), int(vm.available), int(vm.used), float(vm.percent),
        int(vm.free), getattr(vm, 'active', None), getattr(vm, 'inactive', None),
        getattr(vm, 'cached', None), getattr(vm, 'wired', None),
        getattr(vm, 'buffers', None), getattr(vm, 'shared', None), getattr(vm, 'slab', None),
        int(swap.total), int(swap.used), int(swap.free), float(swap.percent),
        int(swap.sin), int(swap.sout),
        int(page_total) if page_total else None,
        int(page_used) if page_used else None,
        json.dumps(raw)
    )
    columns = ["ts","hostname","total_bytes","available_bytes","used_bytes","used_percent","free_bytes","active_bytes","inactive_bytes","cached_bytes","wired_bytes","buffers_bytes","shared_bytes","slab_bytes","swap_total_bytes","swap_used_bytes","swap_free_bytes","swap_used_percent","swap_sin","swap_sout","pagefile_total","pagefile_used","raw"]
    # psutil vm may lack some fields on Windows -> convert None handling via raw; ensure ints
    # Some fields missing -> set to None
    def _fix(v):
        return int(v) if isinstance(v, int) else (int(v) if isinstance(v, float) else v)
    # Already built row with None safe
    return columns, [row]

File: go/internal/collectors/memory.go
package collectors

import (
	"encoding/json"
	"time"

	"github.com/shirou/gopsutil/v4/mem"
)

// CollectMemory mirrors collectors.psutil_collectors.collect_memory.
func CollectMemory(hostname string, ts time.Time) (Result, error) {
	vm, _ := mem.VirtualMemory()
	swap, _ := mem.SwapMemory()

	if vm == nil {
		vm = &mem.VirtualMemoryStat{}
	}
	if swap == nil {
		swap = &mem.SwapMemoryStat{}
	}

	raw := map[string]any{
		"virtual": map[string]any{
			"total": vm.Total, "available": vm.Available, "used": vm.Used,
			"used_percent": vm.UsedPercent, "free": vm.Free, "active": vm.Active,
			"inactive": vm.Inactive, "buffers": vm.Buffers, "cached": vm.Cached,
			"shared": vm.Shared, "wired": vm.Wired, "slab": vm.Slab,
		},
		"swap": map[string]any{
			"total": swap.Total, "used": swap.Used, "free": swap.Free,
			"used_percent": swap.UsedPercent, "sin": swap.Sin, "sout": swap.Sout,
		},
	}
	rawJSON, _ := json.Marshal(raw)

	rows := [][]any{{
		ts, hostname,
		int64(vm.Total), int64(vm.Available), int64(vm.Used), float64(vm.UsedPercent),
		int64(vm.Free), int64(vm.Active), int64(vm.Inactive), int64(vm.Cached),
		int64(vm.Wired), int64(vm.Buffers), int64(vm.Shared), int64(vm.Slab),
		int64(swap.Total), int64(swap.Used), int64(swap.Free), float64(swap.UsedPercent),
		int64(swap.Sin), int64(swap.Sout),
		int64(swap.Total), int64(swap.Used),
		json.RawMessage(rawJSON),
	}}

	return Result{
		Table: "monitor.memory",
		Columns: []string{
			"ts", "hostname",
			"total_bytes", "available_bytes", "used_bytes", "used_percent",
			"free_bytes", "active_bytes", "inactive_bytes", "cached_bytes",
			"wired_bytes", "buffers_bytes", "shared_bytes", "slab_bytes",
			"swap_total_bytes", "swap_used_bytes", "swap_free_bytes", "swap_used_percent",
			"swap_sin", "swap_sout", "pagefile_total", "pagefile_used", "raw",
		},
		Rows: rows,
	}, nil
}
